### ISSUE TITLE
Support goto definition in library files

### ISSUE THREAD

**Issue by MikhailArkhipov** [microsoft/python-language-server#840]
Currently we navigate to library, but then `Goto Def` is no longer available.

**Comment by MikhailArkhipov** [comment]
https://github.com/Microsoft/vscode/issues/71418

**Comment by jakebailey** [comment]
This appears to be coming from the extension, rather than VS Code itself. See Microsoft/vscode-python#5132.

**Comment by MikhailArkhipov** [comment]
Modules outside of the workspace are considered libraries and in order to conserve memory file content, parse tree and navigation information is not preserved. This might be made into an option if memory consumption increase is not a concern.

**Comment by MikhailArkhipov** [comment]
`AnalysisOptions` could have an setting at the expense of memory consumption.

**Comment by rvanlaar** [comment]
As I created issue: #1848 that was closed a duplicate of this one.

I do have three things that irk me about how it works now.

Why:
- doesn't the Language server tell me it won't do libraries when it opens the code?
- doesn't the Language server start introspection when I open a library file?
- when I open a workspace with a library file open does the Language server do introspection on said library file?

I couldn't find documentation about this.
For now it seems Jedi should be recommended instead of the Language server.

**Comment by MikhailArkhipov** [comment]
Libraries are analyzed but information is not completely kept in order to conserve memory.  They are analyzed when imported, library doesn't have to be opened. If you open library file it is not re-analyzed. As mentioned above, a setting may be added to keep all information at the expense of memory.

**Comment by rvanlaar** [comment]
I would like for such a flag to be included.

Just to add: When a library file is open from the start, it is analyzed.

Follow these steps to see this in action:
-> Open a library file in a VSCode workspace
-> Close VSCode
-> Open VSCode
-> Library file is analyzed and goto def is supported in that file.

**Comment by jakebailey** [comment]
For those watching this, you may be able to restore this functionality through a pair of settings that will tell the LS to not drop info for libraries from memory:

```
"python.analysis.memory.keepLibraryAst": true,
"python.analysis.memory.keepLibraryLocalVariables": true,
```

But the memory cost of storing this info will likely be significant, as it will be stored for all referenced libraries in the dependency graph for the opened files.

**Comment by MisLink** [comment]
I tried these two settings but language server still can't jump between library files. I have to say the lack of this feature is a little frustrating.

**Comment by jakebailey** [comment]
This will be available in 0.5.37 and above as an optional feature that can be enabled (#1861) by setting:

```json
"python.analysis.memory.keepLibraryAst": true
```

Note that storing this information about libraries is likely to be very memory heavy, hence why we threw this information away in the first place. Your experience may vary.

**Comment by usuyama** [comment]
I just tried @jakebailey 's suggestion and it's working perfectly on my environment (Windows 10 + SSH to Linux Azure VM). It's great that I can just Go To Definition (F12) for library code/functions. Thank you!
